### DISCUSSION TITLE
NumPy v2 support

### DISCUSSION THREAD

**Discussion by agriyakhetarpal** [online-ml/river#1630] [General]
## Description

Hi all, I'm coming via pyodide/pyodide#4925 where we are trying to update the `river` package that gets built in-tree for the Pyodide distribution via Emscripten. I noticed that `river` does not support NumPy 2.0 yet – I'm not aware of the history behind the pin, but I've also seen that Cython is also pinned to <3, which means that building against NumPy v2 as a build-time dependency isn't possible until that pin gets lifted (and this has come up elsewhere as well, such as https://github.com/online-ml/river/issues/1624#issuecomment-2425048326 and https://github.com/conda-forge/river-feedstock/pull/28). Considering that a lot of packages in the ecosystem now have wheels that support NumPy v2 and are built against it (please see https://github.com/numpy/numpy/issues/26191), I've been working on a branch locally to test out these changes, and I'm able to build without much trouble on a macOS M-series device with Cython 3.0.11 and NumPy 2.1.3 with most tests passing. Please let me know if it's alright to put together a pull request, and I'll do so! Thanks!

Additionally, it would be great to get a release with NumPy 2.0 compatible wheels for the convenience of downstream packagers like us, so that we can point to the release's source distribution to build from, apply patches, etc.

## Additional context for readers

1. The migration guides sourced from the NumPy and Cython documentation websites: 
     - [NumPy 2.0 migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide), and 
     - [Migrating from Cython 0.29 to 3.0](https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html).
2. Links to discussions from the PR linked above: https://github.com/pyodide/pyodide/pull/4925#issuecomment-2456640179 and https://github.com/pyodide/pyodide/pull/4925#issuecomment-2472601867
3. Wheels built against NumPy v2 will remain compatible with both NumPy v1.X and NumPy>=2 versions, so this dependency is required only at build time and not at runtime. The vice versa (the state of wheels built against NumPy v1.X being compatible with NumPy 2.X) is not true.

**Comment by MaxHalford** [comment]
Hey @gbolmier, could you give a hand with this? I think that we can dump `RollingROCAUC` in order to support for Cython 3. And then I don't think upgrading to NumPy v2 will be a big deal.

**Reply by gbolmier** [reply]
Hey @MaxHalford, I'm on a vacation trip until Tuesday... can't guarantee I'll have the time to give a hand on this one

**Comment by MaxHalford** [comment]
Or maybe @e10e3 if you have some time on your hands?

**Reply by e10e3** [reply]
I quickly tried upgrading NumPy, I saw no major hurdle, confirming @agriyakhetarpal's experience. Along with NumPy, Scikit-learn needs to be updated (the first compatible version is 1.4.2).
Even building with Cython 3, I saw no problem regarding RollingROCAUC (I compiled on macOS).

The only thing that needs to be adapted are the tests, NumPy's formatting has changed and causes 21 tests to fail (the underlying values are the same).
The failures look like this:
```
Expected:
    0.994168...
Got:
    np.float64(0.9941683623419704)
```

However, adapting the tests mean this is not just a build-time change.

**Comment by agriyakhetarpal** [comment]
If I can receive a green signal, I can open a PR and keep access to my branch enabled for maintainers to push changes to (in the case where I'm unable to finish it) or pick up my changes from. I don't see anything troubling with `RollingROCAUC` _so far_ that I feel can't be fixed with some revisions to the packaging infrastructure, since it can't find the headers but should be able to compile successfully if it does.

**Reply by gbolmier** [reply]
I think that would be helpful @agriyakhetarpal, don't hesitate to do so :)

**Reply by agriyakhetarpal** [reply]
Awesome! Here we go: https://github.com/online-ml/river/pull/1632
